### PR TITLE
make CoreShop PHP-STAN level 2 compatible

### PR DESCRIPTION
**Pull request by dpfaffenbauer**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | not sure
| Deprecations? | no

### CODE REVIEW

File: src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php [hunk 1]
```diff
@@ -50,14 +50,18 @@ public function __construct(string $type, string $routeName, UrlGeneratorInterfa
...
-         * @var $object Concrete
+         * @var Concrete $object
          */
         Assert::isInstanceOf($object, Concrete::class);
 
         $locale = isset($params['_locale']) ? $params['_locale'] : null;
 
         $name = InheritanceHelper::useInheritedValues(function () use ($object, $locale) {
-            return $object->getName($locale);
+            if (method_exists($object, 'getName')) {
```
[the remark this answers is not in the document]
Comment (dkarlovi) [reply src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php]: Usually means you're missing an interface.

File: src/CoreShop/Bundle/PimcoreBundle/CoreExtension/EmbeddedClass.php [hunk 2]
```diff
@@ -134,12 +134,23 @@ public function getDataFromEditmode($data, $object = null, $params = [])
                     }
 
                     if (method_exists($fd, 'isRemoteOwner') and $fd->isRemoteOwner()) {
-                        $remoteClass = DataObject\ClassDefinition::getByName($fd->getOwnerClassName());
-                        $relations = $embeddedObject->getRelationData($fd->getOwnerFieldName(), false, $remoteClass->getId());
-                        $toAdd = $this->detectAddedRemoteOwnerRelations($relations, $value);
-                        $toDelete = $this->detectDeletedRemoteOwnerRelations($relations, $value);
-                        if (count($toAdd) > 0 or count($toDelete) > 0) {
-                            $this->processRemoteOwnerRelations($embeddedObject, $toDelete, $toAdd, $fd->getOwnerFieldName());
+                        if (method_exists($fd, 'getOwnerClassName') && method_exists($fd, 'getOwnerFieldName')) {
```
[the remark this answers is not in the document]
Comment (dkarlovi) [reply src/CoreShop/Bundle/PimcoreBundle/CoreExtension/EmbeddedClass.php]: > - Well, HE started it!
> - DID NOT!
> - DID TO!

:)

File: src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php]: I know :), but introducing a new Interface will probably not solve the original issue. Why did we introduce the getName here without checking if its available 🤔

File: src/CoreShop/Bundle/PimcoreBundle/CoreExtension/EmbeddedClass.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply src/CoreShop/Bundle/PimcoreBundle/CoreExtension/EmbeddedClass.php]: What I wanted to point out here is: This shouldn't be here at all, but there was no other way to solve it then to copy it. And now I have to make PHPStan happy

File: src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkarlovi) [reply src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php]: Maybe you need to typehint the closure?

File: src/CoreShop/Component/Order/Model/Cart.php [hunk 3]
```diff
@@ -29,7 +29,7 @@ class Cart extends AbstractProposal implements CartInterface
     /**
      * {@inheritdoc}
      */
-    public function getItemForProduct(StorageListProductInterface $product)
+    public function getItemForProduct($product)
```
Comment (solverat) [inline src/CoreShop/Component/Order/Model/Cart.php]: Removing StorageListInterface will throw an exception on existing usages.

File: src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpfaffenbauer) [reply src/CoreShop/Bundle/CoreBundle/Pimcore/LinkGenerator/DataObjectLinkGenerator.php]: No, the problem here is a different one, the LinkGenerator just assumes that there is a `getName` function without even knowing it is there. we just leave this as is and look at that later for BC reasons.

**Pull request merged**